File: job_store.py
"""Disk-backed job store for font generation jobs.

Generation takes minutes — far longer than proxies keep a request open
(Cloudflare tunnels cut it at ~100s). POST /api/generate-font only enqueues
and returns a job_id; clients poll GET /api/job/{id} and download from
GET /api/job/{id}/result. State lives on disk (status.json per job) so it
survives --reload restarts and works with multiple workers.
"""

import json
import logging
import re
import shutil
import tempfile
import threading
import time
from pathlib import Path

logger = logging.getLogger("png2font-api")

JOBS_ROOT = Path(tempfile.gettempdir()) / "png2font_jobs"
JOB_TTL_SECONDS = 2 * 60 * 60  # keep results around for 2 hours
JOB_ID_RE = re.compile(r"^[a-f0-9]{32}$")
HEARTBEAT_SECONDS = 15
# A processing job whose status hasn't been touched for this long has a dead
# worker thread (heartbeats refresh updated_at every HEARTBEAT_SECONDS).
ORPHAN_STALE_SECONDS = 90

# Serializes read-modify-write of status.json between the pipeline thread and
# its heartbeat thread, so a heartbeat can never resurrect a terminal status.
_status_write_lock = threading.Lock()


def job_dir(job_id: str) -> Path:
    return JOBS_ROOT / job_id


def write_job_status(job_id: str, **fields):
    d = job_dir(job_id)
    d.mkdir(parents=True, exist_ok=True)
    status_path = d / "status.json"
    with _status_write_lock:
        current = read_job_status(job_id) or {}
        current.update(fields)
        current["updated_at"] = time.time()
        tmp = status_path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(current))
        tmp.replace(status_path)


def read_job_status(job_id: str):
    try:
        return json.loads((job_dir(job_id) / "status.json").read_text())
    except (FileNotFoundError, json.JSONDecodeError):
        return None


def sweep_stale_jobs():
    """Delete job workspaces older than the TTL. Called on each new submission."""
    if not JOBS_ROOT.exists():
        return
    cutoff = time.time() - JOB_TTL_SECONDS
    for d in JOBS_ROOT.iterdir():
        try:
            if d.is_dir() and d.stat().st_mtime < cutoff:
                shutil.rmtree(d, ignore_errors=True)
                logger.info(f"Swept stale job workspace: {d.name}")
        except OSError:
            pass


def fail_orphaned_jobs():
    """Mark jobs whose worker thread died (server restart) as failed.

    A live worker heartbeats status.json every HEARTBEAT_SECONDS, so a
    non-terminal job with a stale updated_at has no thread behind it.
    Called at startup and lazily from the status endpoint."""
    if not JOBS_ROOT.exists():
        return
    cutoff = time.time() - ORPHAN_STALE_SECONDS
    for d in JOBS_ROOT.iterdir():
        if not d.is_dir():
            continue
        status = read_job_status(d.name)
        if (
            status
            and status.get("status") in ("queued", "processing")
            and status.get("updated_at", 0) < cutoff
        ):
            write_job_status(
                d.name,
                status="failed",
                phase="error",
                detail="The font server restarted during generation — please export again.",
            )
            logger.warning(f"Marked orphaned job as failed: {d.name}")

File: src/jobStore.ts
/**
 * Disk-backed job store — ported from Python job_store.py
 * Stores job status in status.json per job directory, with atomic writes.
 */

import { existsSync, mkdirSync, readFileSync, writeFileSync, renameSync, readdirSync, statSync, rmSync } from "fs";
import { join } from "path";
import { JOBS_ROOT, JOB_TTL_SECONDS, ORPHAN_STALE_SECONDS, JOB_ID_RE } from "./constants.js";
import { JobStatus, JobLifecycleStatus } from "./types.js";

// Per-job lock: Map<jobId, Promise> to serialize concurrent writes (heartbeat + main pipeline)
const perJobLocks = new Map<string, Promise<void>>();

/**
 * Gets the job directory path.
 */
export function jobDir(jobId: string): string {
  return join(JOBS_ROOT, jobId);
}

/**
 * Creates a promise-chain lock for a specific job to serialize writes.
 */
function getJobLock(jobId: string): Promise<void> {
  const existing = perJobLocks.get(jobId) ?? Promise.resolve();
  return existing;
}

/**
 * Acquires and holds a lock for exclusive access to a job's status.json.
 * Returns a function to release the lock.
 */
function acquireJobLock(jobId: string): () => void {
  let releaseLock: () => void = () => {};

  const lockPromise = getJobLock(jobId).then(
    () =>
      new Promise<void>((resolve) => {
        releaseLock = resolve;
      })
  );

  perJobLocks.set(jobId, lockPromise);
  return releaseLock;
}

/**
 * Writes job status to status.json, merging fields (not replacing).
 * Atomic write via temp file + rename, using synchronous fs calls.
 * Serialized per-job to avoid heartbeat callback interleaving with main pipeline writes.
 */
export function writeJobStatus(jobId: string, fields: Partial<JobStatus>): void {
  const release = acquireJobLock(jobId);

  try {
    const dir = jobDir(jobId);
    mkdirSync(dir, { recursive: true });

    const statusPath = join(dir, "status.json");
    const current = readJobStatus(jobId) ?? { status: "queued" as JobLifecycleStatus };

    const updated = {
      ...current,
      ...fields,
      updated_at: Math.floor(Date.now() / 1000), // epoch seconds
    } as JobStatus;

    const tmpPath = `${statusPath}.tmp`;
    writeFileSync(tmpPath, JSON.stringify(updated));
    renameSync(tmpPath, statusPath);
  } finally {
    release();
  }
}

/**
 * Reads job status from status.json.
 * Returns null if the file doesn't exist or can't be parsed.
 */
export function readJobStatus(jobId: string): JobStatus | null {
  try {
    const statusPath = join(jobDir(jobId), "status.json");
    const content = readFileSync(statusPath, "utf-8");
    return JSON.parse(content);
  } catch {
    return null;
  }
}

/**
 * Deletes job workspaces older than the TTL.
 * Called on every new job submission.
 */
export function sweepStaleJobs(): void {
  if (!existsSync(JOBS_ROOT)) {
    return;
  }

  const cutoffTime = Math.floor(Date.now() / 1000) - JOB_TTL_SECONDS;

  try {
    const entries = readdirSync(JOBS_ROOT);
    for (const entry of entries) {
      const path = join(JOBS_ROOT, entry);
      try {
        const stat = statSync(path);
        if (stat.isDirectory() && stat.mtimeMs / 1000 < cutoffTime) {
          rmSync(path, { recursive: true });
        }
      } catch {
        // Ignore errors for individual entries (e.g. permission denied)
      }
    }
  } catch {
    // Ignore errors iterating JOBS_ROOT
  }
}

/**
 * Marks jobs with stale heartbeats (dead worker threads) as failed.
 * A live worker heartbeats updated_at every HEARTBEAT_SECONDS,
 * so non-terminal jobs with stale updated_at have no thread behind them.
 * Called at startup and inline in status checks.
 */
export function failOrphanedJobs(): void {
  if (!existsSync(JOBS_ROOT)) {
    return;
  }

  const cutoffTime = Math.floor(Date.now() / 1000) - ORPHAN_STALE_SECONDS;

  try {
    const entries = readdirSync(JOBS_ROOT);
    for (const jobId of entries) {
      if (!JOB_ID_RE.test(jobId)) {
        continue;
      }

      const status = readJobStatus(jobId);
      if (
        status &&
        (status.status === "queued" || status.status === "processing") &&
        (status.updated_at ?? 0) < cutoffTime
      ) {
        writeJobStatus(jobId, {
          status: "failed",
          phase: "error",
          detail: "The font server restarted during generation — please export again.",
        });
      }
    }
  } catch {
    // Ignore errors
  }
}
